MatrixRTC: introduce custom StateKey type for call member state events.

toger5 opened the pull request:
Initially we have run checks in the matrix-rust-sdk if the key is of correct format. But the key itself was a plain string. This moves that logic to ruma, so that the parsing itself will fail if the format of the state key is not of the appropriate type.

It stores the key as the raw key and a enum with the Owned User Id and Device Id.
(Duplicated memory that is only kept in sync by constructing it correctly)
```
pub struct CallMemberStateKey {
    key: CallMemberStateKeyEnum,
    raw: String,
}
```
This is required because a state key is expected to implement the `AsRef<str>` Trait.
To me it looks like the correct solution would be that CallMemberStateKeyEnum only stores references to the string slices that store the `&UserId` and `&DeviceId` but self-referential types are tricky in rust. a `Rc` type would probably work?
<!--

PR checklist, not strictly necessary but generally useful unless you're just
fixing a typo or something like that:

- Run `cargo xtask ci` locally before posting the PR
- Documented public API changes in CHANGELOG.md files

-->

toger5 committed:
events: introduce custom StateKey type for call member state events.

zecakeh commented:
Maybe we need to change the bound for state keys so that the type must implement `ToString` or `Into<String>` instead…

The only reason I could find that we need this bound is to avoid an allocation on methods like [`AnyStateEvent::state_key()`](https://docs.ruma.dev/ruma/events/enum.AnyStateEvent.html#method.state_key), but I am not sure how much they are used.

zecakeh commented:
Another way to do this is to have a type like [`DeviceKeyId`](https://docs.ruma.dev/ruma/struct.DeviceKeyId.html) or [`SigningKeyId`](https://docs.ruma.dev/ruma/type.SigningKeyId.html), where the inner string is validated during construction/deserialization so we know that we can safely access its parts when we need to.

toger5 commented:
> Another way to do this is to have a type like [DeviceKeyId](https://docs.ruma.dev/ruma/struct.DeviceKeyId.html) or [SigningKeyId](https://docs.ruma.dev/ruma/type.SigningKeyId.html), where the inner string is validated during construction/deserialization so we know that we can safely access its parts when we need to.

this would work but I decided against it since it would imply reparsing/verifying the UserId and splitting the string accordingly each time we request the reference.

> Maybe we need to change the bound for state keys so that the type must implement ToString or Into<String> instead… 

This would be phenomenal. It would allow us to remove the raw representation since we do not need to have the full string in memory somewhere and we could instead construct it using the enum itself.

toger5 commented:
> but I am not sure how much they are used

How much are they actually used? Since we would run to_string also for serializing it, I can imagine, that the additional referencing is happening somewhat often?

We basically have to bite one of the following
 - more memory allocation for call.member events (because the key will store the strings twice)
 - more computation each time we query the user_id/device_id from the key (by running find "_" each time)
 - more memory allocations for serializing for call.member events if we store the key as a combination of `OwnedUserId` and `OwnedDeviceId` and use `ToString` as the trait required for the state_key.
 - more code complexity if we store the index (number) for the separation and don't re-parse the raw state key string but instead just create a `DeviceId from a string slice. (would this require unsafe code if we want to skip the validation step since we already know the underlying string has the correct type?)
 
I think the current version is not too bad it does not impact other event types and is only doubling the initial memory allocation but will not introduce new allocations at other points during program execution.

zecakeh commented on crates/ruma-events/src/state_key.rs:
```diff
@@ -38,3 +39,170 @@ impl Serialize for EmptyStateKey {
+#[allow(clippy::exhaustive_structs)]
+pub struct CallMemberStateKey {
+    key: CallMemberStateKeyEnum,
+    raw: String,
```
Maybe we should use a `Box<str>`, afaik it allocates less space than `String` and is perfect for immutable strings.

zecakeh commented on crates/ruma-events/src/state_key.rs:
```diff
@@ -38,3 +39,170 @@ impl Serialize for EmptyStateKey {
+    where
+        D: Deserializer<'de>,
+    {
+        let s = ruma_common::serde::deserialize_cow_str(deserializer)?;
```
We should add an `impl FromStr for CallMemberStateKey` and use it here.

zecakeh commented on crates/ruma-events/src/state_key.rs:
```diff
@@ -38,3 +39,170 @@ impl Serialize for EmptyStateKey {
+    }
+}
+
+fn parse_rtc_member_event_key(
```
This should be `impl FromStr for CallMemberStateKeyEnum`.

zecakeh commented on crates/ruma-events/src/state_key.rs:
```diff
@@ -38,3 +39,170 @@ impl Serialize for EmptyStateKey {
+            CallMemberStateKeyEnum::UnderscoreUserDevice(u, d) => format!("_{}{}", u, d),
+            CallMemberStateKeyEnum::UserDevice(u, d) => format!("{}{}", u, d),
+            CallMemberStateKeyEnum::User(u) => u.to_string(),
+        };
```
This should be the `impl Display for CallMemberStateKeyEnum`.

We should `impl From<CallMemberStateKeyEnum> for CallMemberStateKey` and call it here instead.

zecakeh commented on crates/ruma-events/src/state_key.rs:
```diff
@@ -1,3 +1,4 @@
+use ruma_common::{DeviceId, OwnedDeviceId, OwnedUserId, UserId};
```
This module is for state key types that are common to several events, this is not the case of the types you added. They should be with the rest of the `m.call.member` types.

zecakeh replied:
This has not been done.

zecakeh commented on crates/ruma-events/src/state_key.rs:
```diff
@@ -38,3 +39,170 @@ impl Serialize for EmptyStateKey {
+
+/// This enum represents all possible formats for a call member event state key.
+#[derive(Clone, Debug, PartialEq, Eq, PartialOrd, Ord)]
+#[allow(clippy::exhaustive_structs)]
```
This is not a struct I also doubt that we want this to be exhaustive for forward-compatibility.

zecakeh commented on crates/ruma-events/src/state_key.rs:
```diff
@@ -38,3 +39,170 @@ impl Serialize for EmptyStateKey {
 }
+
+/// A type that can be used as the `state_key` for call member state events.
+/// Those events can be a combination of UserId and DeviceId.
```
"Those state keys" maybe?

zecakeh reviewed (commented):
I am not sure which representation is the best in the end but let's go with that for now.

Please follow the contributing guide for commit messages, the description should start with `events:`, not `MatrixRTC:`.

toger5 commented:
I force pushed to get the commit message correct.
I implemented all the comments. We now have two ways to get to a CallMemberStateKey. Either we call from_str directly or we call from_str on the CallMemberStateKeyEnum and then into.
`CallMemberStateKey::from_str(key) == CallMemberStateKeyEnum::from_str(key).into()` But behind the scenes they behave different where `CallMemberStateKey::from_str` just stores the `key` in raw and `CallMemberStateKeyEnum::from_str(key).into()` converts into the enum first and then generates the raw manually using `format!`.

I added a test to confirm that this works but I am not sure this is okay.

zecakeh commented on crates/ruma-events/src/state_key.rs:
```diff
@@ -38,3 +39,208 @@ impl Serialize for EmptyStateKey {
+    {
+        let s = ruma_common::serde::deserialize_cow_str(deserializer)?;
+        let string = String::from(s.as_ref());
+        Self::from_str(&string).map_err(|err| de::Error::invalid_value(Unexpected::Str(&s), &err))
```
The first line makes an unnecessary allocation. This should work:

```suggestion
        Self::from_str(&s).map_err(|err| de::Error::invalid_value(Unexpected::Str(&s), &err))
```

zecakeh commented on crates/ruma-events/src/state_key.rs:
```diff
@@ -38,3 +39,208 @@ impl Serialize for EmptyStateKey {
+            CallMemberStateKeyEnum::UserDevice(u, _) => u,
+            CallMemberStateKeyEnum::User(u) => u,
+        }
+        .as_ref()
```
This line should be unnecessary.

zecakeh commented on crates/ruma-events/src/state_key.rs:
```diff
@@ -38,3 +39,208 @@ impl Serialize for EmptyStateKey {
+            CallMemberStateKeyEnum::UserDevice(_, d) => Some(d),
+            CallMemberStateKeyEnum::User(_) => None,
+        }
+        .map(|d| d.as_ref())
```
This line should be unnecessary too.

zecakeh commented on crates/ruma-events/src/state_key.rs:
```diff
@@ -38,3 +39,208 @@ impl Serialize for EmptyStateKey {
+            CallMemberStateKeyEnum::UnderscoreUserDevice(u, d) => format!("_{}_{}", u, d).into(),
+            CallMemberStateKeyEnum::UserDevice(u, d) => format!("{}_{}", u, d).into(),
+            CallMemberStateKeyEnum::User(u) => u.to_string().into(),
+        };
```
As I said in a previous comment, we should implement `std::fmt::Display` for `CallMemberStateKeyEnum` and call it here.

zecakeh commented on crates/ruma-events/src/state_key.rs:
```diff
@@ -38,3 +39,208 @@ impl Serialize for EmptyStateKey {
+enum CallMemberStateKeyEnum {
+    UnderscoreUserDevice(OwnedUserId, OwnedDeviceId),
+    UserDevice(OwnedUserId, OwnedDeviceId),
+    User(OwnedUserId),
```
The variants lack docs.

zecakeh commented on crates/ruma-events/src/call/member/member_state_key.rs:
```diff
@@ -0,0 +1,223 @@
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match &self {
+            CallMemberStateKeyEnum::UnderscoreUserDevice(u, d) => write!(f, "_{}_{}", u, d),
+            CallMemberStateKeyEnum::UserDevice(u, d) => write!(f, "{}_{}", u, d).into(),
```
Prefer named parameters:

```suggestion
            CallMemberStateKeyEnum::UnderscoreUserDevice(u, d) => write!(f, "_{u}_{d}"),
            CallMemberStateKeyEnum::UserDevice(u, d) => write!(f, "{u}_{d}").into(),
```

zecakeh commented on crates/ruma-events/src/call/member/member_state_key.rs:
```diff
@@ -0,0 +1,223 @@
+        match &self {
+            CallMemberStateKeyEnum::UnderscoreUserDevice(u, d) => write!(f, "_{}_{}", u, d),
+            CallMemberStateKeyEnum::UserDevice(u, d) => write!(f, "{}_{}", u, d).into(),
+            CallMemberStateKeyEnum::User(u) => write!(f, "{}", u),
```
In this case this would be cleaner:

```suggestion
            CallMemberStateKeyEnum::User(u) => f.write_str(u.as_str()),
```

zecakeh reviewed (commented):
Just a couple of nits remaining. Have you tested this with the SDK, to make sure that the current public API is what you need?

Also see the error in CI.

toger5 commented:
There is a companion PR in the SDK. It will only work with that.
For now its just a branch ontop of the feat-sss branch.
We should wait for this until the sdk PR is ready.

toger5 commented:
I think we need to merge this before we can get the PR for the rust sdk working.
I am thinking of the following procedure:
 - I build the feat-sss branch with this chagnes locally and check that the rust sdk is working with it.
 - We merge this PR into ruma/main
 - we bump ruma/feat-sss with the change
 - we update the ruma version and adjust the sdk  to use the new api/state key type in one PR for the rust sdk?

toger5 commented:
Here is the draft for the rust-sdk changes. (the test all run locally)
https://github.com/matrix-org/matrix-rust-sdk/pull/3972

We of course need to wait until we can update the cargo.toml to use the correct ruma version.

zecakeh commented on crates/ruma-events/src/call/member.rs:
```diff
@@ -630,4 +639,54 @@ mod tests {
+
+        let parse_result = from_json_value::<AnyStateEvent>(member_event_json(user_id));
+        // let a = format!("Failed parsing: {:?}", parse_result);
+        // assert!(parse_result.is_ok(), "{}", a);
```
It looks like these lines can be removed

zecakeh commented on crates/ruma-events/src/call/member/member_state_key.rs:
```diff
@@ -0,0 +1,225 @@
+    }
+}
+
+/// Error when ttrying to parse a call member state key.
```
```suggestion
/// Error when trying to parse a call member state key.
```

zecakeh commented on crates/ruma-events/src/call/member/member_state_key.rs:
```diff
@@ -0,0 +1,225 @@
+        let state_key: CallMemberStateKey = state_key_enum.into();
+        // This compares state_key.raw (generated) with key (original)
+        assert_eq!(state_key.as_ref(), key);
+        // Campare to the from string without `CallMemberStateKeyEnum`` step.
```
```suggestion
        // Compare to the from string without `CallMemberStateKeyEnum` step.
```

zecakeh reviewed (commented):
Just a bit of cleanup left. Could you rebase on main? I am pretty sure that a new clippy lint will be triggered by your doc comments, so I'd like to catch it now rather than after this is merged.

toger5 commented:
@zecakeh I have done the rebase.

zecakeh reviewed (approved):
Nice!

Hywan commented on crates/ruma-events/src/call/member/member_state_key.rs:
```diff
@@ -0,0 +1,225 @@
+
+impl CallMemberStateKey {
+    /// Constructs a new CallMemberStateKey there are three possible formats:
+    /// - "_{UserId}_{DeviceId}" example: "_@test:user.org_DEVICE". `device_id`: Some`, `underscore:
```
```diff
- `device_id`: Some`
+ `device_id: Some`
```

Hywan replied:
and elsewhere below
